File: src/popup/utils/validation.ts
// Validation utilities for wallet operations

import { PasswordRequirements, MnemonicValidation } from '../types/wallet';
import * as bip39 from 'bip39';

// Complete BIP39 word list for validation
// Using the bip39 library for accurate validation

/**
 * Validates password strength according to security requirements
 */
export function validatePassword(password: string): PasswordRequirements {
  const requirements: PasswordRequirements = {
    minLength: password.length >= 8,
    hasUppercase: /[A-Z]/.test(password),
    hasLowercase: /[a-z]/.test(password),
    hasNumber: /\d/.test(password),
    hasSpecialChar: /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password),
  };

  return requirements;
}

/**
 * Checks if password meets all requirements
 */
export function isPasswordValid(password: string): boolean {
  const requirements = validatePassword(password);
  return Object.values(requirements).every(Boolean);
}

/**
 * Validates mnemonic phrase according to BIP39 standard
 */
export function validateMnemonic(mnemonic: string): MnemonicValidation {
  const words = mnemonic.trim().toLowerCase().split(/\s+/);
  const errors: string[] = [];

  // Check word count
  if (words.length !== 12 && words.length !== 24) {
    errors.push('Mnemonic must be 12 or 24 words');
  }

  // Check if all words are valid BIP39 words using the complete word list
  const invalidWords = words.filter(word => !bip39.wordlists.EN.includes(word));
  if (invalidWords.length > 0) {
    errors.push(`Invalid words: ${invalidWords.join(', ')}`);
  }

  // Check for empty words
  if (words.some(word => word === '')) {
    errors.push('Mnemonic contains empty words');
  }

  // Note: BIP39 allows duplicate words as long as the checksum is valid
  // The actual validation is done in the native backend which has a complete BIP39 implementation

  return {
    isValid: errors.length === 0,
    wordCount: words.length,
    errors,
  };
}

/**
 * Validates chain name
 */
export function validateChain(chain: string): boolean {
  const supportedChains = ['ethereum', 'bsc'];
  return supportedChains.includes(chain.toLowerCase());
}

/**
 * Validates derivation path format
 */
export function validateDerivationPath(path: string): boolean {
  // Basic validation for derivation path format
  const pathRegex = /^m(\/\d+'?)*$/;
  return pathRegex.test(path);
}

/**
 * Sanitizes mnemonic input by normalizing spaces and case
 */
export function sanitizeMnemonic(mnemonic: string): string {
  return mnemonic
    .trim()
    .toLowerCase()
    .replace(/\s+/g, ' ');
}

/**
 * Formats mnemonic for display (with proper spacing)
 */
export function formatMnemonicForDisplay(mnemonic: string): string {
  const words = sanitizeMnemonic(mnemonic).split(' ');
  const formatted: string[] = [];
  
  for (let i = 0; i < words.length; i += 4) {
    formatted.push(words.slice(i, i + 4).join(' '));
  }
  
  return formatted.join('\n');
}

/**
 * Generates a readable password strength message
 */
export function getPasswordStrengthMessage(password: string): string {
  const requirements = validatePassword(password);
  const failedRequirements: string[] = [];

  if (!requirements.minLength) failedRequirements.push('at least 8 characters');
  if (!requirements.hasUppercase) failedRequirements.push('an uppercase letter');
  if (!requirements.hasLowercase) failedRequirements.push('a lowercase letter');
  if (!requirements.hasNumber) failedRequirements.push('a number');
  if (!requirements.hasSpecialChar) failedRequirements.push('a special character');

  if (failedRequirements.length === 0) {
    return 'Password meets all requirements';
  }

  return `Password must contain: ${failedRequirements.join(', ')}`;
}

File: native/pkg/wallet/validation.go
// SPDX-License-Identifier: Apache-2.0
package wallet

import (
	"errors"
	"fmt"
	"strings"

	bip39 "github.com/tyler-smith/go-bip39"
)

// ValidateMnemonic validates a BIP39 mnemonic phrase
func ValidateMnemonic(mnemonic string) error {
	if mnemonic == "" {
		return errors.New("mnemonic phrase cannot be empty")
	}

	// Normalize the mnemonic (trim spaces, lowercase)
	normalizedMnemonic := strings.TrimSpace(strings.ToLower(mnemonic))
	
	// Check word count (BIP39 supports 12, 15, 18, 21, 24 words)
	words := strings.Fields(normalizedMnemonic)
	wordCount := len(words)
	
	validWordCounts := []int{12, 15, 18, 21, 24}
	isValidCount := false
	for _, count := range validWordCounts {
		if wordCount == count {
			isValidCount = true
			break
		}
	}
	
	if !isValidCount {
		return fmt.Errorf("invalid mnemonic word count: %d (must be 12, 15, 18, 21, or 24 words)", wordCount)
	}

	// Check for duplicate words (common issue that makes BIP39 validation fail)
	wordMap := make(map[string][]int)
	for i, word := range words {
		wordMap[word] = append(wordMap[word], i+1) // 1-based position for user-friendly error
	}
	
	var duplicates []string
	for word, positions := range wordMap {
		if len(positions) > 1 {
			duplicates = append(duplicates, fmt.Sprintf("'%s' (positions %v)", word, positions))
		}
	}
	
	if len(duplicates) > 0 {
		return fmt.Errorf("invalid mnemonic: duplicate words found: %s", strings.Join(duplicates, ", "))
	}
	
	// Check if the mnemonic is valid according to BIP39
	if !bip39.IsMnemonicValid(normalizedMnemonic) {
		return errors.New("invalid mnemonic phrase")
	}

	return nil
}

// ValidatePassword performs basic password validation
func ValidatePassword(password string) error {
	if password == "" {
		return errors.New("password cannot be empty")
	}

	if len(password) < 8 {
		return errors.New("password must be at least 8 characters long")
	}

	// Additional password strength checks could be added here
	// For now, we keep it simple but functional

	return nil
}

// ValidateChain validates if the chain is supported for wallet import
func ValidateChain(chain string) error {
	if chain == "" {
		return errors.New("chain cannot be empty")
	}

	// Normalize chain name
	normalizedChain := strings.ToLower(strings.TrimSpace(chain))
	
	supportedChains := []string{"ethereum", "eth", "bsc", "binance"}
	for _, supported := range supportedChains {
		if normalizedChain == supported {
			return nil
		}
	}

	return fmt.Errorf("unsupported chain: %s (supported: ethereum, bsc)", chain)
}

// NormalizeChain normalizes chain names to standard format
func NormalizeChain(chain string) string {
	normalizedChain := strings.ToLower(strings.TrimSpace(chain))
	
	switch normalizedChain {
	case "eth", "ethereum":
		return "ethereum"
	case "bsc", "binance":
		return "bsc"
	default:
		return normalizedChain
	}
}